### ISSUE TITLE
[BUG] clientCertificates troubleshooting

### ISSUE THREAD

**Issue by mxschmitt** [microsoft/playwright#32004]
> Update: This issue should be fixed in the latest [@canary](https://playwright.dev/docs/canary-releases). See #31906 for further discussions.

Hi @mxschmitt I have tried this with the beta release and testOptions but it doesn't works as expected, I am still getting the pop-up asking to select the certificate.

![image](https://github.com/user-attachments/assets/755803de-e046-4f8e-8760-d9fefd11b4ef)

here's my config:
![image](https://github.com/user-attachments/assets/35cef9f4-6b22-4943-b37d-a811f8f3f8bf)

package.json:
![image](https://github.com/user-attachments/assets/1d8cb3be-3dee-49f1-b6ac-6afb917775fa)

test file:
![image](https://github.com/user-attachments/assets/77bc5ebc-30c0-47c9-ba41-70b203abf517)

_Originally posted by @amitAutoS in https://github.com/microsoft/playwright/issues/1799#issuecomment-2268097809_

**Comment by mxschmitt** [comment]
@amitAutoS posting it here, so that we don't send notifications to the others. I see you are on Windows, are you using Google Chrome by any chance?

Do you mind providing us your full playwright.config.ts?

**Comment by amitAutoS** [comment]
> @amitAutoS posting it here, so that we don't send notifications to the others. I see you are on Windows, are you using Google Chrome by any chance?
> 
> Do you mind providing us your full playwright.config.ts?

sure, here it is:

import { defineConfig, devices } from "@playwright/test";
import { resolve } from "path";
import { readFileSync } from "fs";

/**
 * See https://playwright.dev/docs/test-configuration.
 */
// const pfxPaths = resolve(__dirname, "./certs/DavidJonesTest_G5.pfx");
// console.log("pfxPaths: " + pfxPaths);
export default defineConfig({
  testDir: "./e2e",
  /* Run tests in files in parallel */
  fullyParallel: true,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: [["list"], ["html", { open: "never" }]],
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Base URL to use in actions like `await page.goto('/')`. */
    // baseURL: 'http://127.0.0.1:3000',

    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: "on-first-retry",
    headless: false,
    ignoreHTTPSErrors: true,
    clientCertificates: [
      {
        origin: "https://myurl.com/",
        pfxPath: "C:/Users/myusername/Downloads/client_cert/Test_G5.pfx",
        passphrase: "passwordhere",
      },
    ],
  },

  /* Configure projects for major browsers */
  projects: [
    // {
    //   name: "chromium",
    //   use: {
    //     ...devices["Desktop Chrome"],
    //   },
    // },
    // {
    //   name: "firefox",
    //   use: {
    //     ...devices["Desktop Firefox"],
    //   },
    // },
    // {
    //   name: "webkit",
    //   use: { ...devices["Desktop Safari"] },
    // },
    /* Test against branded browsers. */
    // {
    //   name: 'Microsoft Edge',
    //   use: { ...devices['Desktop Edge'], channel: 'msedge' },
    // },
    {
      name: "Google Chrome",
      use: { ...devices["Desktop Chrome"], channel: "chrome" },
    },
  ],

  /* Run your local dev server before starting the tests */
  // webServer: {
  //   command: 'npm run start',
  //   url: 'http://127.0.0.1:3000',
  //   reuseExistingServer: !process.env.CI,
  // },
});

**Comment by mxschmitt** [comment]
Yes Edge Stable and Chrome Stable have both the same bug. Could you try e.g. just chromium or beta?

**Comment by amitAutoS** [comment]
Tried Chromium with no channel. Same behavior as Firefox. Tried to load the page(I dont see the cert popup) but couldn't load and dies.

**Comment by amitAutoS** [comment]
> launchOptions:

I tried this but getting error: 
Error: browser.newContext: Cannot specify both proxy and clientCertificates

![image](https://github.com/user-attachments/assets/3021a0d3-943b-45d8-a6c2-08fa88b9457b)

**Comment by mxschmitt** [comment]
Could you try to set the `DEBUG=pw:client-certificates` environment variable and rerun it? If this doesn't yield anything helpful, we need to add more debug logs to troubleshoot this issue.

**Comment by amitAutoS** [comment]
getting this error:

![image](https://github.com/user-attachments/assets/7fa2031f-9770-48c0-a911-24b70dc23ca3)

**Comment by mxschmitt** [comment]
So no additional output when you set the envs? Do you mind pinging me on [Discord](https://aka.ms/playwright/discord)? Happy to investigate further there.

**Comment by ArmandGrechScerri** [comment]
I am also encountering the exact same above issue. Has a fix/ workaround been found?

**Comment by mxschmitt** [comment]
> I am also encountering the exact same above issue. Has a fix/ workaround been found?

Please provide us your config file, this would help us a ton in order to triage this issue further.

**Comment by ArmandGrechScerri** [comment]
We use a bit of a custom launcher. Here is the config I am using for this specific issue:

```
import { cpus } from "os";

import type { PlaywrightTestConfig } from "@playwright/test";
import { loadProjects } from "e2e/src/services/playwright.service";

// Generate unique output folder to store old local test run reports
const currentDateTime = new Date().toISOString().replace(/[:.]/g, "_").slice(0, -1);
const localOutputFolder = `../test-reports/obgaj-end-to-end-${currentDateTime}`;

const ciOutputFolder = `../test-reports`;

/**
 * See https://playwright.dev/docs/test-configuration.
 */
const config: PlaywrightTestConfig = {
	testDir: `../../../apps/${process.env.PWC_DOMAIN ?? "adaptive"}/end-to-end`,
	/* Maximum time one test can run for. */
	timeout: 180_000,
	expect: {
		/**
		 * Maximum time expect() should wait for the condition to be met.
		 * For example in `await expect(locator).toHaveText();`
		 */
		timeout: 20_000,
	},
	/* Run tests in files in parallel */
	fullyParallel: true,
	/* Retry on CI only */
	retries: 1,
	/* Opt out of parallel tests on CI. */
	workers: process.env.CI ? 3 : Math.floor(cpus().length * 0.35),
	/* Reporter to use. See https://playwright.dev/docs/test-reporters */
	reporter: process.env.CI
		? [["blob", { outputDir: ciOutputFolder }]]
		: [["list"], ["html", { outputFolder: localOutputFolder }], ["../src/custom-reporters/pass-rate-reporter.ts"]],
	/* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
	use: {
		headless: true,
		viewport: { width: 1280, height: 720 },
		ignoreHTTPSErrors: true,
		video: process.env.CI ? "off" : "retain-on-failure",
		trace: process.env.CI ? "off" : "on-first-retry",
		screenshot: "only-on-failure",
		navigationTimeout: 30_000,
		testIdAttribute: "data-test-id",
		actionTimeout: 30_000,
		clientCertificates: [
			{
				origin: "*ORIGIN URL*",
				certPath: "../certs/cert.pem",
				keyPath: "../certs/key.pem",
				passphrase: "*PASSPHRASE*",
			},
		],
	},
	testMatch: /.*e2e\.(js|ts|mjs)/,
	projects: loadProjects(),
};

export default config;
```

**Comment by mxschmitt** [comment]
Feel free to ping me on Discord. `loadProjects` would be the interesting piece.

**Comment by mxschmitt** [comment]
I'm going to close this one for now since I believe the issues in this issue were addressed. We recommend 

    npm install -D @playwright/test@beta

until we have released v1.46.1. Thank you both for your time to debug these issues together. For further reference:

- https://github.com/microsoft/playwright/pull/32155 - fixes an issue with Microsoft IIS
- https://github.com/microsoft/playwright/pull/32163 - fixes an issue when TLS handshake fails (e.g. using TLS certificates which OpenSSL considers legacy)
